### DISCUSSION TITLE
Reading glb with external resource uri (with absolute paths)

### DISCUSSION THREAD

**Discussion by repalash** [donmccurdy/glTF-Transform#644] [Q&A] [answered]
Hi, 
I am trying to read GLB with some external URI images and write them with Draco compression.

 I see from https://github.com/donmccurdy/glTF-Transform/issues/193#issuecomment-785329983 that it's not possible to resolve relative paths in `binaryToJSON`, but absolute paths should be possible.

In the source of `binaryToJSON` it throws an error when any URI is found. 

Is there any way to handle this? I thought of replacing the external images with placeholder data URIs, but I guess from the source, that will not work either.

Thanks.

**Comment by donmccurdy** [comment, marked as answer]
If you use a method that takes a URI or path, these external references can be resolved:

```javascript
const document = await io.readBinary('path/to/model.glb');
const jsonDocument = await io.readAsJSON('path/to/model.glb');
```

If the GLB is already in memory and that's not an option, then I think the only workaround the library currently supports would be to unpack the GLB container manually ([example](https://github.com/donmccurdy/glTF-Transform/blob/d710f158a6fd79e7caac571d686819dd5b4acd2d/packages/core/src/io/platform-io.ts#L279-L315)), append the resources you need to the `resources` map, and then parse it with `io.readJSON(...)`.

Perhaps we could make that second method a bit easier by adding a second parameter to the `binaryToJSON` method, allowing you to "opt out" of the checks for external resources:

```javascript
const jsonDocument = io.binaryToJSON(binary, skipExternal = true);
```

That option hasn't been added yet, and isn't the default behavior because I do think it's important to know that the JSONDocument returned is not actually self-contained. But adding the option would sound reasonable to me.

**Reply by repalash** [reply]
I see, didn't realise there is already a readExternalResources function already through the `readAsJSON` path. Thanks for this, I will try it out today.  (Maybe this could be mentioned in the `throw Error` in `binaryToJSON` function?) 
For glb, `readBinary` doesn't take a string argument, I guess you mean `io.read`.
It looks like passing the data URL should work for memory objects.

But, would it work for object URLs created from memory objects with `URL.createObjectURL`? appending the extension with `#.glb` .

Adding a `skipExternal` parameter to `binaryToJSON` sounds great, or to make the functions `_binaryToJSON` and `_readResourcesInternal/External` public

Thanks

**Reply by repalash** [reply]
I see `readResourceExternal` also downloads the file with `fetch`, is it also possible to skip the external downloads and work with just the URI?

**Reply by donmccurdy** [reply]
> I see readResourceExternal also downloads the file with fetch, is it also possible to skip the external downloads  ...

I think that'd be what the `skipExternal` flag should do – would you want to open a feature request ticket for this?
